### PR TITLE
issue-487 NBS-4829 [disk manager]: make possible to add arbitrary logging field to context, write fields from context in ydb client logger

### PR DESCRIPTION
**Pull request by gy2411**
Added methods that allow to add logging fields to context. Used that methods to improve ydb client logging: now it writes field from context.

Examples of rows after these changes:
```
Feb 27 13:00:29 d9hnumvcrda87ck5kbpm disk-manager[305304]: ydb_logger.go:73 DEBUG => COMPONENT: YDB, YDB_LOG_NAME: ydb.driver.resolver.init, ENDPOINT: localhost:32648, DATABASE: /Root, SECURE: false => start
Feb 27 13:00:29 d9hnumvcrda87ck5kbpm disk-manager[305304]: factory.go:69 DEBUG => calling client.Close for &{0xc0005c2480 300000000000 500000000 0x17f5fe0 0xc0004b0260}, localhost:17106
```
(in 'real' logs there also will be request id, idempotency key and operation id fiedls)

### CODE REVIEW

File: cloud/tasks/logging/journald_logger.go [hunk 1]
```diff
@@ -42,23 +47,104 @@ func getPriority(level Level) journal.Priority {
...
+		rawField := getField(f)
+		vars[rawField.key] = rawField.value
 	}
 
 	return vars
 }
 
+// Takes the latest value for each key
+// Preserves order of fields
+func deduplicateFields(
```
Comment (gy2411) [inline cloud/tasks/logging/journald_logger.go:82]: Для записи полей в сообщение важно дедуплицировать поля, сохранив их порядок, чтобы в аналогичных друг другу строчках лога поля шли в одном и том же порядке.

File: cloud/tasks/logging/journald_logger.go [hunk 2]
```diff
@@ -88,35 +174,40 @@ func (l *journaldLogger) structuredLogWithCaller(
 	msg string,
 	fields ...log.Field,
 ) {
+	ff := newJournaldLoggerFileds(fields...)
+
+	if len(ff.messageFields) > 0 {
+		msg = serializeFields(ff.messageFields) + " => " + msg
```
Comment (gy2411) [inline cloud/tasks/logging/journald_logger.go:180]: Error "=>" appears around the fields and also around the message body. Like this:
Prefix => fields => message body => error.

File: cloud/tasks/logging/journald_logger.go [hunk 3]
```diff
@@ -42,23 +47,104 @@ func getPriority(level Level) journal.Priority {
...
+}
+
+func newJournaldLoggerFileds(fields ...log.Field) (ff journaldLoggerFileds) {
+	fmt.Printf("newJournaldLoggerFileds: len = %d, %v\n", len(fields), fields)
+	ff.journaldFields = make(map[string]string)
+
+	var messageFieldsAll []rawField
+
+	for _, f := range fields {
+		if f.Key() == syslogIdentifierKey {
```
Comment (gy2411) [inline cloud/tasks/logging/journald_logger.go:127]: We need to choose which fields to wriite to journald fields (would not be seen in journalctl without -o flag) and which fields to write to the message itself.

Currently we write syslog identifier, idempotency key, request id and operation id to journald fields.

In this pr syslog identifier is written to journald fields, and all other fields are written to the message.

It it ok?

File: cloud/tasks/logging/fields.go [hunk 4]
```diff
@@ -49,25 +53,51 @@ func Any(key string, value interface{}) Field {
...
+	return String("IMAGE_ID", value)
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+func WithFields(ctx context.Context, fields ...Field) context.Context {
+	return ctxlog.WithFields(ctx, fields...)
+}
+
+func WithGeneralFields(ctx context.Context) context.Context {
```
Comment (jkuradobery) [inline cloud/tasks/logging/fields.go:82]: WithCommonFields

File: cloud/tasks/logging/journald_logger.go [hunk 5]
```diff
@@ -42,23 +47,100 @@ func getPriority(level Level) journal.Priority {
 
 ////////////////////////////////////////////////////////////////////////////////
 
-func getFields(fields ...log.Field) map[string]string {
+type rawField struct {
```
Comment (jkuradobery) [inline cloud/tasks/logging/journald_logger.go:50]: I do not see how is the extraction of the getField() method helps here? I do not see any other usage of the method. I think we should keep the code as it was working previously.

Comment (jkuradobery) [reply cloud/tasks/logging/journald_logger.go:50]: Sorry, didn't notice the usage of the rawField

File: cloud/tasks/logging/journald_logger.go (hunk 3 above)
Comment (jkuradobery) [reply cloud/tasks/logging/journald_logger.go:126]: I think we can duplicate them. journalctl would allow us to perform filtering and tracing.

**Review by jkuradobery — APPROVED** [review]

**Review by leftmain — APPROVED** [review]

**Pull request merged**